gsoc-2024/ideas.md

```diff
@@ -0,0 +1,211 @@
+The project length can be varied depending on which of these items are covered by the project.
+
+#### Reading & Related Material
+
+* https://www.musl-libc.org/
+* https://unikraft.org/guides/using-the-app-catalog
+* https://github.com/unikraft/catalog
+* https://unikraft.org/docs/contributing/adding-to-the-app-catalog
+
+---
+
+### Software Quality Assurance of Unikraft Codebase
+
+| | |
+|-|-|
+| **Difficulty** | 3/5 |
+| **Project Size** | Variable (175 or 350 hours) |
+| **Maximum instances** | 2 |
+| **Constraints/requirements** | C programming skills, Linux command-line experience, build tools |
+
+#### Description
+
+During its 5 years of existence, Unikraft, now at version 0.12, has grown in features, application support and codebase.
+As it matures, a high quality of the code and robust behavior are a must to provide a stable solution for its user base.
+
+The aim of this project is to assist in the software quality assurance of the Unikraft codebase, by tackling one of the following ideas:
+
+1. The use of the [`uktest` framework](https://github.com/unikraft/unikraft/tree/staging/lib/uktest) to create unit tests for [internal libraries](https://github.com/unikraft/unikraft/tree/staging/lib/) and [external libraries](https://github.com/search?q=topic%3Alibrary+org%3Aunikraft+fork%3Atrue&type=repositories).
+   Not many libraries have unit tests, those that do are mostly exceptions.
+   This will directly impact the stability of the code base and allow quick validation of new features that may break existing functionality.
+
+1. Inclusion of static and dynamic analysis tools that highlight potential spots of faulty or undefined behavior.
+
+1. The use of compiler builtins and compiler flags providing constraints on the code to increase its resilience to faulty behavior.
+
+1. Augmenting the CI/CD system used by Unikraft (based on [GitHub Actions](https://github.com/features/actions)) to feature automatic testing, validation and vetting of contributions to Unikraft repositories: core, libraries, applications.
+   Potential items are:
+   1. handling running of unikernels instead of simple builds
+   1. static analysis of images to be delivered as reports to GitHub pull requests
+   1. regression checks on performance (delivered as % change from the current upstream version)
+
+Any other project that is targeted towards increasing the robustness of Unikraft source code is welcome.
+These will both increase the viability of Unikraft as a stable solution and increase the quality of future contributions, by enforcing good practices on submitted code.
+
+#### Reading & Related Material
+
+* [Writing Tests in Unikraft](https://unikraft.org/docs/develop/writing-tests/)
+* https://www.guru99.com/unit-testing-guide.html
+* https://docs.kernel.org/dev-tools/kunit/index.html
+* https://github.com/features/actions
+* https://unikraft.org/docs/contributing/review-process/
+
+---
+
+### Driver Probing Framework
```

Comment:
This goes beyond what GSoC time would afford... I suggest we handle this ourselves.